### PR TITLE
ShellPkg/Acpiview: Add HEST Parser

### PR DESCRIPTION
**Pull request by LeviYeoReum**
# Description

Add a new pareser for the Hardware Error Source Table (HEST).

The HEST table is used to describe a system's hardware error sources to OSPM.

Cc: Zhichao Gao <zhichao.gao@intel.com>
Cc: Sami Mujawar <sami.mujawar@arm.com>
Cc: Pierre Gondois <pierre.gondois@arm.com>

### CODE REVIEW

**Comment by LeviYeoReum** [comment]
Hi, @LiuZhiguang001.
Could your review this patch series please? Thanks!

File: ShellPkg/Library/UefiShellAcpiViewCommandLib/Parsers/Hest/HestParser.c [hunk 1]
```diff
@@ -0,0 +1,958 @@
...
+}
+
+/**
+  Helper macro to populate the header fields of error source descriptor in the
+  ACPI_PARSER array.
+**/
+#define PARSE_HEST_ERROR_SOURCE_COMMON_HEADER(FlagsValidateFunc, CheckRecordCount)   \
+  { L"Type", 2, 0, L"%d", NULL, NULL, NULL, NULL },                           \
+  { L"Source Id", 2, 2, L"%d", NULL, NULL, NULL, NULL },                      \
+  { L"Reserved", 2, 4, NULL, NULL, NULL, NULL, NULL },                     \
```
Comment (pierregondois) [inline ShellPkg/Library/UefiShellAcpiViewCommandLib/Parsers/Hest/HestParser.c:542]: Uncrustify doesn't seem to complain about this,
but would it be possible to align the ending '\' of the macros (similar comments for macros below)

Comment (LeviYeoReum) [reply ShellPkg/Library/UefiShellAcpiViewCommandLib/Parsers/Hest/HestParser.c]: Thanks! I'll fix it :)

**Review by pierregondois — APPROVED** [review]

**Comment by pierregondois** [comment]
Hi @ZhichaoGao , 
Just a ping in case it was forgotten

**Comment by LeviYeoReum** [comment]
Hi @ZhichaoGao 
Just a ping in case it was forgotten

**Comment by pierregondois** [comment]
Hi @lgao4,
@ZhichaoGao doesn't seem to be available. Would it be ok to merge this patch as you merged the EINJ parser ?

Regards,
Pierre

**Comment by lgao4** [comment]
> Hi @lgao4, @ZhichaoGao doesn't seem to be available. Would it be ok to merge this patch as you merged the EINJ parser ?
> 
> Regards, Pierre

Can you attach EINJ table dump result? Is it the expected result?

**Comment by LeviYeoReum** [comment]
Hi @lgao4.

Here is EINJ && Hest Table print example
```
 --------------- EINJ Table --------------- 

Address  : 0xF97DF598
Length   : 240
  
00000000 : 45 49 4E 4A F0 00 00 00 - 01 71 41 52 4D 4C 54 44   EINJ.....qARMLTD
00000010 : 41 52 4D 53 47 49 20 20 - 27 07 14 20 41 52 4D 20   ARMSGI  '.. ARM 
00000020 : 99 00 00 00 30 00 00 00 - 00 00 00 00 06 00 00 00   ....0...........
00000030 : 03 03 00 00 00 20 00 03 - 00 00 61 FF 00 00 00 00   ..... ....a.....
00000040 : 07 00 00 00 00 00 00 00 - FF FF FF FF FF FF FF FF   ................
00000050 : 05 03 00 00 00 40 00 04 - 00 00 61 FF 00 00 00 00   .....@....a.....
00000060 : 00 00 00 00 00 00 00 00 - FF FF FF FF FF FF FF FF   ................
00000070 : 08 02 00 00 00 40 00 04 - 00 01 61 FF 00 00 00 00   .....@....a.....
00000080 : 00 00 00 00 00 00 00 00 - FF FF FF FF FF FF FF FF   ................
00000090 : 06 00 00 00 00 40 00 04 - 50 02 61 FF 00 00 00 00   .....@..P.a.....
000000A0 : 00 00 00 00 00 00 00 00 - FF FF FF FF FF FF FF FF   ................
000000B0 : 07 00 00 00 00 40 00 04 - 00 04 61 FF 00 00 00 00   .....@....a.....
000000C0 : 00 00 00 00 00 00 00 00 - FF FF FF FF FF FF FF FF   ................
000000D0 : 01 03 00 00 00 40 00 04 - 00 02 61 FF 00 00 00 00   .....@....a.....
000000E0 : 00 10 61 FF 00 00 00 00 - FF FF FF FF FF FF FF FF   ..a.............

Table Checksum : OK

EINJ                                 :
  Signature                          : EINJ
  Length                             : 240
  Revision                           : 1
  Checksum                           : 0x71
  Oem ID                             : ARMLTD
  Oem Table ID                       : ARMSGI  
  Oem Revision                       : 0x20140727
  Creator ID                         : ARM 
  Creator Revision                   : 0x99
  Injection Header Size              : 48
  Injection Flags                    : 0x0
  Reserved                           : 
  Injection Entry Count              : 6
  Injection Instruction Entry        :
    Injection Action                 : GET_ERROR_TYPE(0x3)
    Instruction                      : WRITE_REGISTER_VALUE(0x3)
    Flags                            : 0x0
    Reserved                         : 
    Register Region                  : 
      Address Space ID               : 0x0
      Register Bit Width             : 0x20
      Register Bit Offset            : 0x0
      Access Size                    : 0x3
      Address                        : 0xFF610000

    Value                            : 0x7
    Mask                             : 0xFFFFFFFF
  Injection Instruction Entry        :
    Injection Action                 : EXECUTE_OPERATION(0x5)
    Instruction                      : WRITE_REGISTER_VALUE(0x3)
    Flags                            : 0x0
    Reserved                         : 
    Register Region                  : 
      Address Space ID               : 0x0
      Register Bit Width             : 0x40
      Register Bit Offset            : 0x0
      Access Size                    : 0x4
      Address                        : 0xFF610000

    Value                            : 0x0
    Mask                             : 0xFFFFFFFF
  Injection Instruction Entry        :
    Injection Action                 : SET_ERROR_TYPE_WITH_ADDRESS(0x8)
    Instruction                      : WRITE_REGISTER(0x2)
    Flags                            : 0x0
    Reserved                         : 
    Register Region                  : 
      Address Space ID               : 0x0
      Register Bit Width             : 0x40
      Register Bit Offset            : 0x0
      Access Size                    : 0x4
      Address                        : 0xFF610100

    Value                            : 0x0
    Mask                             : 0xFFFFFFFF
  Injection Instruction Entry        :
    Injection Action                 : CHECK_BUSY_STATUS(0x6)
    Instruction                      : READ_REGISTER(0x0)
    Flags                            : 0x0
    Reserved                         : 
    Register Region                  : 
      Address Space ID               : 0x0
      Register Bit Width             : 0x40
      Register Bit Offset            : 0x0
      Access Size                    : 0x4
      Address                        : 0xFF610250

    Value                            : 0x0
    Mask                             : 0xFFFFFFFF
  Injection Instruction Entry        :
    Injection Action                 : GET_COMMAND_STATUS(0x7)
    Instruction                      : READ_REGISTER(0x0)
    Flags                            : 0x0
    Reserved                         : 
    Register Region                  : 
      Address Space ID               : 0x0
      Register Bit Width             : 0x40
      Register Bit Offset            : 0x0
      Access Size                    : 0x4
      Address                        : 0xFF610400

    Value                            : 0x0
    Mask                             : 0xFFFFFFFF
  Injection Instruction Entry        :
    Injection Action                 : GET_TRIGGER_ERROR_ACTION_TABLE(0x1)
    Instruction                      : WRITE_REGISTER_VALUE(0x3)
    Flags                            : 0x0
    Reserved                         : 
    Register Region                  : 
      Address Space ID               : 0x0
      Register Bit Width             : 0x40
      Register Bit Offset            : 0x0
      Access Size                    : 0x4
      Address                        : 0xFF610200

    Value                            : 0xFF611000
    Mask                             : 0xFFFFFFFF

Table Statistics:
        0 Error(s)
```

```
 --------------- HEST Table --------------- 

Address  : 0xF97DE818
Length   : 224
  
00000000 : 48 45 53 54 E0 00 00 00 - 01 6B 41 52 4D 4C 54 44   HEST.....kARMLTD
00000010 : 52 45 46 49 4E 46 52 41 - 31 08 20 20 41 52 4D 20   REFINFRA1.  ARM 
00000020 : 01 00 00 00 02 00 00 00 - 0A 00 0A 00 FF FF 00 01   ................
00000030 : 01 00 00 00 01 00 00 00 - 60 03 00 00 00 40 00 04   ........`....@..
00000040 : 08 10 62 FF 00 00 00 00 - 0B 1C 00 00 00 00 00 00   ..b.............
00000050 : 25 03 00 00 00 00 00 00 - 00 00 00 00 00 00 00 00   %...............
00000060 : 00 00 00 00 BC 03 00 00 - 00 40 00 04 00 10 62 FF   .........@....b.
00000070 : 00 00 00 00 00 00 00 00 - 00 00 00 00 00 00 00 00   ................
00000080 : 00 00 00 00 0A 00 14 00 - FF FF 00 01 01 00 00 00   ................
00000090 : 01 00 00 00 50 00 00 00 - 00 40 00 04 08 20 62 FF   ....P....@... b.
000000A0 : 00 00 00 00 0B 1C 00 00 - 00 00 00 00 24 03 00 00   ............$...
000000B0 : 00 00 00 00 00 00 00 00 - 00 00 00 00 00 00 00 00   ................
000000C0 : AC 00 00 00 00 40 00 04 - 00 20 62 FF 00 00 00 00   .....@... b.....
000000D0 : 00 00 00 00 00 00 00 00 - 00 00 00 00 00 00 00 00   ................

Table Checksum : OK

HEST                                 :
  Signature                          : HEST
  Length                             : 224
  Revision                           : 1
  Checksum                           : 0x6B
  Oem ID                             : ARMLTD
  Oem Table ID                       : REFINFRA
  Oem Revision                       : 0x20200831
  Creator ID                         : ARM 
  Creator Revision                   : 0x1
  Error Source Count                 : 2
  Generic Hardware Error Source V2 Structure :
    Type                             : 10
    Source Id                        : 10
    Related Source Id                : 0xFFFF
    Flags                            : 0x0
    Enabled                          : 1
    Number of Records to Pre-allocate : 1
    Max Sections Per Record          : 1
    Max Raw Data Length              : 864
    Error Status Address             : 
      Address Space ID               : 0x0
      Register Bit Width             : 0x40
      Register Bit Offset            : 0x0
      Access Size                    : 0x4
      Address                        : 0xFF621008

    Notification Structure           : 
      Type                           : Software Delegated Exception(0xB)
      Length                         : 28
      Configuration Write Enable     : 0x0
        Type                         : 0
        Poll Interval                : 0
        Switch To Polling Threshold Value : 0
        Switch To Polling Threshold Window : 0
        Error Threshold Value        : 0
        Error Threshold Window       : 0

      Pull Interval                  : 0 ms
      Vector                         : 805
      Switch To Polling Threshold Value : 0
      Switch To Polling Threshold Window : 0 ms
      Error Threshold Value          : 0
      Error Threshold Window         : 0 ms

    Error Status Block Length        : 956
    Read Ack Register                : 
      Address Space ID               : 0x0
      Register Bit Width             : 0x40
      Register Bit Offset            : 0x0
      Access Size                    : 0x4
      Address                        : 0xFF621000

    Read Ack Preserve                : 0
    Read Ack Write                   : 0
  Generic Hardware Error Source V2 Structure :
    Type                             : 10
    Source Id                        : 20
    Related Source Id                : 0xFFFF
    Flags                            : 0x0
    Enabled                          : 1
    Number of Records to Pre-allocate : 1
    Max Sections Per Record          : 1
    Max Raw Data Length              : 80
    Error Status Address             : 
      Address Space ID               : 0x0
      Register Bit Width             : 0x40
      Register Bit Offset            : 0x0
      Access Size                    : 0x4
      Address                        : 0xFF622008

    Notification Structure           : 
      Type                           : Software Delegated Exception(0xB)
      Length                         : 28
      Configuration Write Enable     : 0x0
        Type                         : 0
        Poll Interval                : 0
        Switch To Polling Threshold Value : 0
        Switch To Polling Threshold Window : 0
        Error Threshold Value        : 0
        Error Threshold Window       : 0

      Pull Interval                  : 0 ms
      Vector                         : 804
      Switch To Polling Threshold Value : 0
      Switch To Polling Threshold Window : 0 ms
      Error Threshold Value          : 0
      Error Threshold Window         : 0 ms

    Error Status Block Length        : 172
    Read Ack Register                : 
      Address Space ID               : 0x0
      Register Bit Width             : 0x40
      Register Bit Offset            : 0x0
      Access Size                    : 0x4
      Address                        : 0xFF622000

    Read Ack Preserve                : 0
    Read Ack Write                   : 0

Table Statistics:
        0 Error(s)
        0 Warning(s)

```

**Review by lgao4 — APPROVED** [review]

**Pull request merged**